### PR TITLE
[coq_makefile] Fail when no `CONF_FILE` is present.

### PR DESCRIPTION
**Pull request by ejgallego**
When a `Makefile.conf` file is not present, `coq_makefile` used to try
to re-generate it. This results in #8731.

Instead, we require that the `.conf` file is present, or else we fail;
this means that users cannot distribute generated makefiles
anymore. Indeed, this makes sense as `coq_makefile` writes hardcoded
paths into makefiles.

### CODE REVIEW

File: tools/CoqMakefile.in [hunk 1]
```diff
@@ -33,7 +33,13 @@ COQLIBS           := $(COQMF_COQLIBS)
 COQLIBS_NOML      := $(COQMF_COQLIBS_NOML)
 CMDLINE_COQLIBS   := $(COQMF_CMDLINE_COQLIBS)
 LOCAL             := $(COQMF_LOCAL)
-COQLIB            := $(COQMF_COQLIB)
+# We add a workaround for #8731; if COQLIB is set externally, don't
+# override it.
```
Comment (gares) [inline tools/CoqMakefile.in]: This is not what the line below does. It takes coqlib from a coqtop available in the path whenever COQMF_COQLIB is empty (I've no idea when this is the case).

**Comment by gares** [comment]
Please add a test to the test suite

File: tools/CoqMakefile.in (hunk 1 above)
Comment (ejgallego) [reply tools/CoqMakefile.in]: The case happens as described in the bug report.

Indeed I am far from sure this is the right thing to do here.

Comment (SkySkimmer) [reply tools/CoqMakefile.in]: The point is that the comment doesn't match the code.

File: tools/CoqMakefile.in [hunk 2]
```diff
@@ -33,7 +33,13 @@ COQLIBS           := $(COQMF_COQLIBS)
...
+ifeq ($(COQMF_COQLIB),)
+   COQLIB := $(shell coqtop -where)/
```
Comment (SkySkimmer) [inline tools/CoqMakefile.in]: No need for / at the end, that's a COQBIN thing

File: tools/CoqMakefile.in (hunk 1 above)
Comment (SkySkimmer) [reply tools/CoqMakefile.in]: That is, it sounds like we preserve environment COQLIB unconditionally but we only do so when COQMF_COQLIB is unset.

File: lib/envars.ml [hunk 3]
```diff
@@ -12,7 +12,13 @@ open Util
...
 (** {1 Helper functions} *)
 
-let getenv_else s dft = try Sys.getenv s with Not_found -> dft ()
+let getenv_else s dft =
+  try
+    let v = Sys.getenv s in
+    if String.equal v ""
+    then dft ()
+    else v
+  with Not_found -> dft ()
```
Comment (SkySkimmer) [inline lib/envars.ml]: Why is this needed?

Comment (ejgallego) [reply lib/envars.ml]: I was doing as suggested here https://github.com/coq/coq/issues/8731#issuecomment-429842201

Comment (ejgallego) [reply lib/envars.ml]: So indeed it makes Coq to act with the dummy setting of COQLIB in the makefile as if it was not set. Pretty hacky tho.

Comment (SkySkimmer) [reply lib/envars.ml]: Then why is the other change needed?

Comment (ejgallego) [reply lib/envars.ml]: Likely it broke some other stuff? I wrote this patch long time ago so trying to remember...

**Review by gares — CHANGES_REQUESTED** [review]
Please remove the first hunk since we don't know what it is supposed to solve (and the comment is wrong, that is not a good sign).

**Comment by gares** [comment]
I've restarted the job on iris, it seemed a spurious failure.
The please, before force-pushing, write down if it worked, so that we know if the first hunk did serve any purpose (in practice).

File: lib/envars.ml (hunk 3 above)
Comment (SkySkimmer) [reply lib/envars.ml]: I think what happened is that the first hunk was suggested first, but it's insufficient as COQLIB still gets overridden in the makefile so we would get the default even when the user set it in the environment.
Then the second hunk was written and made the first one not needed.

**Comment by ejgallego** [comment]
> Please remove the first hunk since we don't know what it is supposed to solve (and the comment is wrong, that is not a good sign).

@gares the "first hunk" I see is the one in `envars` but I think you mean to remove the one in `CoqMakefile.in` right?

**Comment by SkySkimmer** [comment]
No, remove the envars one.

**Comment by gares** [comment]
> remove the one in `CoqMakefile.in` right?

This is what I meant.

Actually with the fix in envars you don't need the conditional in the makefile, since an empty value is OK.


Also, the current code calls coqtop from the path,  while the rest of the Makefile honors for example COQBIN, so I think it is introducing a new bug.

**Comment by ejgallego** [comment]
Ok will reshuffle again; but I should introduce a test to see if this is working or not, as the bug is hard to trigger.

**Comment by ejgallego** [comment]
It seems to me that the bug is really in the makefile tho; this kind of self-regeneration cannot bring anything good.

**Comment by ejgallego** [comment]
Ok so I could finally recall why we need two fixes. The real bug is that `CoqMakefile.in` is broken when there is no `Makefile.conf` and it decides to truncate the main makefile...

So, we could handle this by adding two steps:
- fix CoqMakefile.in so when people regenerate their makefiles this doesn't happen again,
- however, this means that some projects still distribute the broken Makefile , That's why we need the fix in `envars`.

However as pointed out by Enrico, it seems that projects are distributing bugged `Makefiles` [without the Makefile.conf], but indeed they cannot distribute that files as it contains absolute paths.

Should we instead just have `CoqMakefile.in` fail if `Makefile.conf` is not there? It makes no sense that `Makefile` doesn't error and instead it silently fails if you don't have a required file.

**Comment by ejgallego** [comment]
> Should we instead just have `CoqMakefile.in` fail if `Makefile.conf` is not there?

Ok, I've implemented this; a couple of CI contribs will need fixing, but seems like the cleaner option.

**Comment by ejgallego** [comment]
Ping @gares , this is a completely new version.

**Comment by gares** [comment]
This removes a featurethat has been there since coqmf 1.

**Comment by ejgallego** [comment]
> This removes a feature that has been there since coq_makefile 1.

Umm, I am not sure I follow, this forbids to distribute a makefile without the corresponding `.conf`, which you said is broken; [and indeed it is]. What is exactly the feature here?

So what should be done then?

**Comment by gares** [comment]
The feature is that if you edit the _CoqProject file then the Makefile (and the .conf) are re-generated.

**Comment by ejgallego** [comment]
I see, I dunno; I'm afraid I am unable to fix this problem.

File: tools/CoqMakefile.in [hunk 4]
```diff
@@ -39,9 +39,6 @@ OCAMLFIND         := $(COQMF_OCAMLFIND)
 CAMLFLAGS         := $(COQMF_CAMLFLAGS)
 HASNATDYNLINK     := $(COQMF_HASNATDYNLINK)
 
-@CONF_FILE@: @PROJECT_FILE@
-	@COQ_MAKEFILE_INVOCATION@
```
Comment (SkySkimmer) [inline tools/CoqMakefile.in]: You could just add `[ - f @CONF_FILE@ ]` (maybe + some nice error message) before the invocation, then if the conf file is present it can be updated but if it's missing you get an error.

Comment (SkySkimmer) [reply tools/CoqMakefile.in]: Or something using makefile's `wildcard` function
